pkg/chaosdaemon/tasks/pod_container_handler.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2022 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tasks
+
+import (
+	"sync"
+
+	"github.com/go-logr/logr"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/cerr"
+)
+
+var ErrNotPodContainerName = cerr.NotType[PodContainerName]()
+
+var ErrPodProcessMapNotInit = cerr.NotInit[map[PodContainerName]SysPID]().WrapName("PodContainerNameProcessMap").Err()
+
+type PodContainerName string
+
+func (p PodContainerName) ToID() string {
+	return string(p)
+}
+
+// ChaosOnPOD stand for the inner process injector for container.
+type ChaosOnPOD interface {
+	Injectable
+	Recoverable
+}
+
+type PodContainerNameProcessMap struct {
+	m      map[PodContainerName]SysPID
+	rwLock sync.RWMutex
+}
+
+func NewPodProcessMap() PodContainerNameProcessMap {
+	return PodContainerNameProcessMap{
+		m:      make(map[PodContainerName]SysPID),
+		rwLock: sync.RWMutex{},
+	}
+}
+
+func (p *PodContainerNameProcessMap) Read(PodContainerName PodContainerName) (SysPID, error) {
+	p.rwLock.RLock()
+	defer p.rwLock.RUnlock()
+	sysPID, ok := p.m[PodContainerName]
+	if !ok {
+		return SysPID(0), ErrNotFoundSysID.WithStack().Err()
+	}
+	return sysPID, nil
+}
+
+func (p *PodContainerNameProcessMap) Write(PodContainerName PodContainerName, sysPID SysPID) {
```

Comment:
I don't understand why this struct is necessary? This maps the "podcontainername" to a pid, but the pid can also be got from `containerId`. Whenever I got a podcontainername, the containerId is usually also touchable, doesn't it?

Reply:
But the `containerId` is not a stable key, if the container restart `containerId` will change. In chaos-daemon we need a stable ID to mark the process tasks applid on.
If use  `containerId` : container restart  and `containerId` changed from "0" to "1", we need apply all the tasks with `containerId` again to refresh the tasks.

Reply:
I just don't understand why we need to store it. For any "Apply" and "Recover", we can get the pid directly, without reading from any existing map, so why we need to read it from the map (by `podcontainername`)?

Reply:
If one of process restart with different PID, we need to know which PID is the last one cause we need to apply tasks from the last process to the new process.The `map[podcontainername]` is for merging tasks.

Reply:
On another hand, I choose store something for merging tasks apply on a same process.

Reply:
> I choose store something for merging tasks apply on a same process

It seems that the `TaskConfigMap` is used to merge the tasks, and this map is not read during the merge :thinking: . (Please correct me if I was wrong).

Reply:
`podcontainername` is also stored in `TaskConfigMap`.

Reply:
`podcontainername` is the value of `TaskConfigMap` & key of `taskmap`. `PodContainerNameProcessMap`  store the relationship of `podcontainername` and working process PID.

Reply:
> If one of process restart with different PID, we need to know which PID is the last one cause we need to apply tasks from the last process to the new process.The `map[podcontainername]` is for merging tasks.

Everytime we call `Apply` or `Recover`, we actually know the pid.

Reply:
Yes . And I do not know the problem now.

Reply:
After discussing with @Andrewmatilde, I understand the problem.

1. The map is used to pass the `pid` to the `PodHandler`, so that the `PodHandler` knows which process to inject.
2. Why don't we pass the `pid` as an argument to the `PodHandler`? Because we cannot replace the argument of `Inject` to the `pid` (or `container_id`), they are not stable.

Every K-V in this map works like a global variable, to tell the `PodHandler` pid. I'm thinking whether there are better way (like using closure or something else) to pass the information to `PodHandler`. (At least, if the abstraction doesn't fit in our logic, we should refractor the abstraction, but not use a "global variable" to bypass it).

Reply:
But actually the map is not a `global variable`  and we can control the shared scope. I think we can add pid to time-chaos-config and use config.Assign & config.New to transfer the pid, but I think it is a dirtier choise.